File: ios/Plugin/Socket/Socket.swift
import Network

class Socket {
    private static let INPUT_STREAM_BUFFER_SIZE = 65536
    
    public let uuid: SocketUuid
    private let options: SocketOptions
    private var delegate: SocketDelegate
    private let connection: NWConnection
    private var openConnectionContinuation: CheckedContinuation<(), Error>? = nil
    private var disconnectCheckingContinuation: CheckedContinuation<(), Error>? = nil
    private var state: SocketState = SocketState.Initial
    
    init(uuid: SocketUuid, options: SocketOptions, delegate: SocketDelegate) {
        self.uuid = uuid
        self.options = options
        self.delegate = delegate
        self.connection = Socket.createConnection(options)
        connection.stateUpdateHandler = stateDidChange(to:)
    }
    
    func open() async throws {
        return try await withCheckedThrowingContinuation { continuation in
            openConnectionContinuation = continuation
            setupReceive()
            connection.start(queue: .main)
        }
    }
    
    func disconnect() async throws {
        return try await withCheckedThrowingContinuation {continuation in
            disconnectCheckingContinuation = continuation
            state = .Closed
            connection.cancel()
        }
    }
    
    func send(_ data: Array<UInt8>) async throws {
        return try await withCheckedThrowingContinuation {continuation in
            connection.send(content: Data(data), completion: NWConnection.SendCompletion.contentProcessed { error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume()
                }
            })
        }
    }
    
    private func setupReceive() {
        connection.receive(minimumIncompleteLength: 1, maximumLength: Socket.INPUT_STREAM_BUFFER_SIZE) { [self] (data, _, isComplete, error) in
            if let data = data, !data.isEmpty {
                onData(data)
            }
            if let error = error {
                onError(error)
            } else if isComplete {
                onClose()
            } else {
                self.setupReceive()
            }
        }
    }
    
    private func stateDidChange(to state: NWConnection.State) {
        switch state {
        case .preparing:
            break
        case .waiting(let error):
            onError(error)
            break
        case .ready:
            onReady()
            break
        case .cancelled:
            onCancelled()
            break
        default:
            break
        }
    }
    
    private func onData(_ data: Data) {
        let bytes = data.map { $0 }
        delegate.onData(socket: self, data: bytes)
    }
    
    private func onClose() {
        if (self.state == .Closed) {
            return
        }
        if (self.state == .Error) {
            return
        }
        delegate.onClose(socket: self)
        self.state = .Closed
        disposeInternal()
    }
    
    private func onReady() {
        if let continuation = openConnectionContinuation {
            openConnectionContinuation = nil
            continuation.resume()
            self.state = .Opened
        }
    }
    
    private func onCancelled() {
        if let continuation = disconnectCheckingContinuation {
            disconnectCheckingContinuation = nil
            continuation.resume()
            onClose()
        }
    }
    
    private func onError(_ error: Error) {
        if (self.state == .Closed) {
            return
        }
        if (self.state == .Error) {
            return
        }
        self.state = .Error
        if let continuation = openConnectionContinuation {
            openConnectionContinuation = nil
            continuation.resume(throwing: error)
        } else {
            delegate.onError(socket: self, error: error)
        }
        disposeInternal()
    }
    
    private func disposeInternal() {
        connection.cancel()
    }
    
    private static func createConnection(_ options: SocketOptions) -> NWConnection {
        let host = NWEndpoint.Host(options.host)
        let port = NWEndpoint.Port(rawValue: UInt16(options.port))!
        return NWConnection(host: host, port: port, using: .tcp)
    }
}

File: android/src/main/java/com/spryrocks/capplugin/socketconnection/socket/Socket.kt
package com.spryrocks.capplugin.socketconnection.socket

import com.spryrocks.capplugin.socketconnection.SocketData
import com.spryrocks.capplugin.socketconnection.SocketUuid
import java.net.InetSocketAddress
import java.net.Socket
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine
import java.net.Socket as JavaSocket

class Socket(
    val uuid: SocketUuid,
    private val options: SocketOptions,
    private val delegate: SocketDelegate
) {
    private val INPUT_STREAM_BUFFER_SIZE = 16 * 1024

    private val connection = createConnection(options)

    private val _stateLock = Object()
    private var state: SocketState = SocketState.Initial

    suspend fun open() {
        suspendCoroutine {
            executor.execute {
                try {
                    val host = options.host
                    val port = options.port
                    connection.connect(InetSocketAddress(host, port))
                    setupReceive()
                    it.resume(Unit)
                    synchronized(_stateLock) {
                        state = SocketState.Opened
                    }
                } catch (error: Throwable) {
                    it.resumeWithException(error)
                    synchronized(_stateLock) {
                        state = SocketState.Error
                    }
                    disposeInternal()
                }
            }
        }
    }

    suspend fun close() {
        suspendCoroutine {
            executor.execute {
                try {
                    connection.close()
                    it.resume(Unit)
                } catch (error: Throwable) {
                    it.resumeWithException(error)
                } finally {
                    synchronized(_stateLock) {
                        state = SocketState.Closed
                    }
                }
            }
        }
    }

    suspend fun send(data: SocketData) {
        suspendCoroutine {
            executor.execute {
                try {
                    connection.getOutputStream().write(data)
                    it.resume(Unit)
                } catch (error: Throwable) {
                    it.resumeWithException(error)
                }
            }
        }
    }

    private fun setupReceive() {
        executor.execute {
            try {
                val buffer = ByteArray(INPUT_STREAM_BUFFER_SIZE)
                var bytesRead: Int

                while (connection.getInputStream().read(buffer).also { bytesRead = it } >= 0) {
                    onData(buffer.copyOfRange(0, bytesRead))
                }
            } catch (error: Throwable) {
                onError(error)
            } finally {
                onClose()
            }
        }
    }

    private fun onData(data: SocketData) {
        delegate.onData(this, data)
    }

    private fun onClose() {
        synchronized(_stateLock) {
            if (state == SocketState.Closed) return
            if (state == SocketState.Error) return
        }
        delegate.onClose(this)
        synchronized(_stateLock) {
            state = SocketState.Closed
        }
        disposeInternal()
    }

    private fun onError(error: Throwable) {
        synchronized(_stateLock) {
            if (state == SocketState.Closed) return
            if (state == SocketState.Error) return
        }
        delegate.onError(this, error)
        synchronized(_stateLock) {
            state = SocketState.Error
        }
        disposeInternal()
    }

    private fun disposeInternal() {
        executor.execute {
            try {
                connection.close()
            } catch (ignored: Throwable) {
            }
        }
    }

    private val executor get() = delegate.executor

    companion object {
        private fun createConnection(options: SocketOptions): Socket {
            return JavaSocket().apply {
                soTimeout = 0
            }
        }
    }
}
